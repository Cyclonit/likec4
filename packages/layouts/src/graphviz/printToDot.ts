/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Colors, RelationColors, compareFqnHierarchically } from '@likec4/core'
import type { ComputedNode, ComputedView, Fqn } from '@likec4/core/types'
import { groupBy, sortObject, values } from 'rambdax'
import {
  attribute as _,
  digraph,
  toDot,
  type GraphBaseModel,
  type NodeModel,
  type SubgraphModel
} from 'ts-graphviz'
import type { DotSource } from './graphviz-types'
import { generateEdgeLabel, generateNodeLabel, pxToInch, pxToPoints } from './graphviz-utils'

export function printToDot({ autoLayout, nodes, edges }: ComputedView): DotSource {
  const gvSubgraphs = new Map<Fqn, SubgraphModel>()
  const gvNodes = new Map<Fqn, NodeModel>()

  let sequence = 1

  const processNode = (node: ComputedNode, parent: GraphBaseModel) => {
    if (node.children.length > 0) {
      const subgraph = parent.createSubgraph('cluster_' + sequence++, {
        [_.id]: node.id,
        [_.labeljust]: 'l',
        [_.fontsize]: pxToPoints(12),
        [_.label]: `<<B>${node.title.toUpperCase()}</B>>`,
        [_.margin]: node.children.length > 2 ? 30 : 20
      })
      gvSubgraphs.set(node.id, subgraph)

      for (const child of nodes.filter(n => n.parent === node.id)) {
        processNode(child, subgraph)
      }
    } else {
      const gNode = parent.createNode('nd' + sequence++, {
        [_.id]: node.id,
        [_.label]: generateNodeLabel(node)
      })
      if (node.color !== 'primary') {
        gNode.attributes.set(_.color, Colors[node.color].stroke)
        gNode.attributes.set(_.fillcolor, Colors[node.color].fill)
      }
      switch (node.shape) {
        case 'cylinder':
        case 'storage': {
          gNode.attributes.set(_.shape, 'cylinder')
          break
        }
        default:
          break
      }
      gvNodes.set(node.id, gNode)
    }
  }

  const G = digraph({
    [_.compound]: true,
    [_.pad]: 0.08,
    [_.rankdir]: autoLayout,
    [_.nodesep]: pxToInch(100),
    [_.ranksep]: pxToInch(100),
    [_.layout]: 'dot',
    [_.TBbalance]: 'min',
    [_.fontname]: 'Helvetica',
    [_.fontsize]: pxToPoints(16)
  })

  G.attributes.node.apply({
    [_.fontname]: 'Helvetica',
    [_.labelloc]: 'c',
    [_.shape]: 'rect',
    [_.width]: pxToInch(320),
    [_.height]: pxToInch(180),
    [_.style]: 'filled,rounded',
    [_.color]: Colors.primary.stroke,
    [_.fillcolor]: Colors.primary.fill,
    [_.margin]: `${pxToInch(32)},${pxToInch(20)}`
  })

  G.attributes.edge.apply({
    [_.fontname]: 'Helvetica',
    [_.style]: 'solid',
    [_.penwidth]: 2,
    [_.arrowsize]: 0.7,
    [_.color]: RelationColors.lineColor,
    [_.fontcolor]: RelationColors.labelColor,
    [_.nojustify]: true
  })

  for (const root of nodes.filter(n => n.parent === null)) {
    processNode(root, G)
  }

  const sortedEdges = values(
    sortObject(
      (parentA, parentB) => compareFqnHierarchically(parentA, parentB),
      groupBy(
        e => e.parent ?? '',
        edges
      )
    )
  )

  for (const edgesPerContainer of sortedEdges) {
    for (const edge of edgesPerContainer) {
      // const container = (edge.parent && gvSubgraphs.get(edge.parent)) ?? G
      const source = gvNodes.get(edge.source)
      const target = gvNodes.get(edge.target)
      if (source && target) {
        const e = G.edge([source, target], {
          [_.id]: edge.id
        })
        const label = generateEdgeLabel(edge)
        if (label) {
          e.attributes.set(_.label, label)
        }
        // this is the only edge in the container
        // and the container has no subgraphs
        // so we can remove the constraint
        if (edgesPerContainer.length === 1) {
          const sourceNd = nodes.find(n => n.id === edge.source)
          const targetNd = nodes.find(n => n.id === edge.target)
          if (sourceNd && targetNd && sourceNd.parent === targetNd.parent) {
            e.attributes.set(_.minlen, 0)
          }
        }
      }
    }
  }

  return toDot(G) as DotSource
}
