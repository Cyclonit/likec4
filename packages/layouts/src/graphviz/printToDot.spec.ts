import { expect, test } from 'vitest'
import {
  computedAmazonView,
  computedCloudView,
  computedCloud3levels,
  computedIndexView
} from '../__mocks__/data'
import { printToDot } from './printToDot'

test('printToDot: computedIndexView', () => {
  const diagram = printToDot(computedIndexView)
  expect(diagram).toMatchSnapshot()
})

test('printToDot: computedAmazonView', () => {
  const diagram = printToDot(computedAmazonView)
  expect(diagram).toMatchSnapshot()
})

test('printToDot: computedCloud3levels', () => {
  const diagram = printToDot(computedCloud3levels)
  expect(diagram).toMatchSnapshot()
})

test('printToDot: computedCloudView', () => {
  const diagram = printToDot(computedCloudView)
  expect(diagram).toMatchSnapshot()
})
