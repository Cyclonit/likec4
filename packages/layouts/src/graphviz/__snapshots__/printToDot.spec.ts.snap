// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`printToDot: computedAmazonView 1`] = `
"digraph {
  compound = true;
  pad = 0.08;
  rankdir = "TB";
  nodesep = 1.0417;
  ranksep = 1.0417;
  layout = "dot";
  TBbalance = "min";
  fontname = "Helvetica";
  fontsize = 12;
  edge [
    fontname = "Helvetica";
    style = "solid";
    penwidth = 2;
    arrowsize = 0.7;
    color = "#a3a3a3";
    fontcolor = "#d4d4d4";
    nojustify = true;
  ];
  node [
    fontname = "Helvetica";
    labelloc = "c";
    shape = "rect";
    width = 3.3333;
    height = 1.875;
    style = "filled,rounded";
    color = "#2563eb";
    fillcolor = "#3b82f6";
    margin = "0.3333,0.2083";
  ];
  subgraph "cluster_1" {
    id = "cloud";
    labeljust = "l";
    fontsize = 9;
    label = <<B>CLOUD</B>>;
    margin = 20;
    "nd2" [
      id = "cloud.backend";
      label = <<FONT color="#eff6ff" point-size="13.5">Backend<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT>>;
    ];
  }
  subgraph "cluster_3" {
    id = "amazon";
    labeljust = "l";
    fontsize = 9;
    label = <<B>AMAZON</B>>;
    margin = 20;
    "nd4" [
      id = "amazon.s3";
      label = <<FONT color="#eff6ff" point-size="13.5">S3<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">S3 is a storage service<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
    ];
  }
  "nd2" -> "nd4" [
    id = "cloud.backend:amazon.s3";
    label = <<FONT point-size="10.5">persists artifacts<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
}"
`;

exports[`printToDot: computedCloud3levels 1`] = `
"digraph {
  compound = true;
  pad = 0.08;
  rankdir = "TB";
  nodesep = 1.0417;
  ranksep = 1.0417;
  layout = "dot";
  TBbalance = "min";
  fontname = "Helvetica";
  fontsize = 12;
  edge [
    fontname = "Helvetica";
    style = "solid";
    penwidth = 2;
    arrowsize = 0.7;
    color = "#a3a3a3";
    fontcolor = "#d4d4d4";
    nojustify = true;
  ];
  node [
    fontname = "Helvetica";
    labelloc = "c";
    shape = "rect";
    width = 3.3333;
    height = 1.875;
    style = "filled,rounded";
    color = "#2563eb";
    fillcolor = "#3b82f6";
    margin = "0.3333,0.2083";
  ];
  "nd1" [
    id = "customer";
    label = <<FONT color="#eff6ff" point-size="13.5">customer<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  "nd2" [
    id = "support";
    label = <<FONT color="#eff6ff" point-size="13.5">Support Engineer<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">Support engineers are responsible for<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT>supporting customers<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  "nd9" [
    id = "amazon";
    label = <<FONT color="#eff6ff" point-size="13.5">Amazon<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">Amazon is a cloud provider<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  subgraph "cluster_3" {
    id = "cloud";
    labeljust = "l";
    fontsize = 9;
    label = <<B>CLOUD</B>>;
    margin = 30;
    "nd4" [
      id = "cloud.frontend.adminPanel";
      label = <<FONT color="#eff6ff" point-size="13.5">Admin Panel Webapp<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">The admin panel is a webapp that allows<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT>support staff to manage customer data.<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
    ];
    "nd5" [
      id = "cloud.frontend.dashboard";
      label = <<FONT color="#eff6ff" point-size="13.5">Customer Dashboard Webapp<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">The customer dashboard is a webapp that<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT>allows customers to view their data.<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
    ];
    subgraph "cluster_6" {
      id = "cloud.backend";
      labeljust = "l";
      fontsize = 9;
      label = <<B>BACKEND</B>>;
      margin = 20;
      "nd7" [
        id = "cloud.backend.graphql";
        label = <<FONT color="#eff6ff" point-size="13.5">Graphql API<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">Component that allows to query data via<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT>GraphQL.<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
      ];
      "nd8" [
        id = "cloud.backend.storage";
        label = <<FONT color="#eff6ff" point-size="13.5">Backend Storage<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">The backend storage is a component that<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT>stores data.<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
        shape = "cylinder";
      ];
    }
  }
  "nd4" -> "nd7" [
    id = "cloud.frontend.adminPanel:cloud.backend.graphql";
    label = <<FONT point-size="10.5">fetches data in zero trust network with<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT>sso authentification<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd5" -> "nd7" [
    id = "cloud.frontend.dashboard:cloud.backend.graphql";
    label = <<FONT point-size="10.5">fetches data<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd8" -> "nd9" [
    id = "cloud.backend.storage:amazon";
    label = <<FONT point-size="10.5">persists artifacts<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd1" -> "nd5" [
    id = "customer:cloud.frontend.dashboard";
    label = <<FONT point-size="10.5">opens in browser<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd2" -> "nd4" [
    id = "support:cloud.frontend.adminPanel";
    label = <<FONT point-size="10.5">manages<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd7" -> "nd8" [
    id = "cloud.backend.graphql:cloud.backend.storage";
    minlen = 0;
  ];
}"
`;

exports[`printToDot: computedCloudView 1`] = `
"digraph {
  compound = true;
  pad = 0.08;
  rankdir = "TB";
  nodesep = 1.0417;
  ranksep = 1.0417;
  layout = "dot";
  TBbalance = "min";
  fontname = "Helvetica";
  fontsize = 12;
  edge [
    fontname = "Helvetica";
    style = "solid";
    penwidth = 2;
    arrowsize = 0.7;
    color = "#a3a3a3";
    fontcolor = "#d4d4d4";
    nojustify = true;
  ];
  node [
    fontname = "Helvetica";
    labelloc = "c";
    shape = "rect";
    width = 3.3333;
    height = 1.875;
    style = "filled,rounded";
    color = "#2563eb";
    fillcolor = "#3b82f6";
    margin = "0.3333,0.2083";
  ];
  "nd1" [
    id = "customer";
    label = <<FONT color="#eff6ff" point-size="13.5">customer<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  "nd2" [
    id = "support";
    label = <<FONT color="#eff6ff" point-size="13.5">Support Engineer<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">Support engineers are responsible for<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT>supporting customers<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  "nd6" [
    id = "amazon";
    label = <<FONT color="#eff6ff" point-size="13.5">Amazon<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">Amazon is a cloud provider<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  subgraph "cluster_3" {
    id = "cloud";
    labeljust = "l";
    fontsize = 9;
    label = <<B>CLOUD</B>>;
    margin = 20;
    "nd4" [
      id = "cloud.frontend";
      label = <<FONT color="#eff6ff" point-size="13.5">Frontend<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT>>;
    ];
    "nd5" [
      id = "cloud.backend";
      label = <<FONT color="#eff6ff" point-size="13.5">Backend<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT>>;
    ];
  }
  "nd4" -> "nd5" [
    id = "cloud.frontend:cloud.backend";
    label = <<FONT point-size="10.5">fetches data<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
    minlen = 0;
  ];
  "nd5" -> "nd6" [
    id = "cloud.backend:amazon";
    label = <<FONT point-size="10.5">persists artifacts<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd1" -> "nd4" [
    id = "customer:cloud.frontend";
    label = <<FONT point-size="10.5">opens in browser<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd2" -> "nd4" [
    id = "support:cloud.frontend";
    label = <<FONT point-size="10.5">manages<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
}"
`;

exports[`printToDot: computedIndexView 1`] = `
"digraph {
  compound = true;
  pad = 0.08;
  rankdir = "TB";
  nodesep = 1.0417;
  ranksep = 1.0417;
  layout = "dot";
  TBbalance = "min";
  fontname = "Helvetica";
  fontsize = 12;
  edge [
    fontname = "Helvetica";
    style = "solid";
    penwidth = 2;
    arrowsize = 0.7;
    color = "#a3a3a3";
    fontcolor = "#d4d4d4";
    nojustify = true;
  ];
  node [
    fontname = "Helvetica";
    labelloc = "c";
    shape = "rect";
    width = 3.3333;
    height = 1.875;
    style = "filled,rounded";
    color = "#2563eb";
    fillcolor = "#3b82f6";
    margin = "0.3333,0.2083";
  ];
  "nd1" [
    id = "customer";
    label = <<FONT color="#eff6ff" point-size="13.5">customer<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  "nd2" [
    id = "support";
    label = <<FONT color="#eff6ff" point-size="13.5">Support Engineer<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">Support engineers are responsible for<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT>supporting customers<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  "nd3" [
    id = "cloud";
    label = <<FONT color="#eff6ff" point-size="13.5">cloud<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  "nd4" [
    id = "amazon";
    label = <<FONT color="#eff6ff" point-size="13.5">Amazon<FONT point-size="15"> <BR ALIGN="CENTER"/></FONT></FONT><FONT point-size="10.5"> <BR/></FONT><FONT color="#dbeafe" point-size="10.5">Amazon is a cloud provider<FONT point-size="12"> <BR ALIGN="CENTER"/></FONT></FONT>>;
  ];
  "nd3" -> "nd4" [
    id = "cloud:amazon";
    label = <<FONT point-size="10.5">persists artifacts<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd1" -> "nd3" [
    id = "customer:cloud";
    label = <<FONT point-size="10.5">opens in browser<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
  "nd2" -> "nd3" [
    id = "support:cloud";
    label = <<FONT point-size="10.5">manages<FONT point-size="12"> <BR ALIGN="LEFT"/></FONT></FONT>>;
  ];
}"
`;
